compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -29,6 +30,54 @@ use provenance_map::*;
 
 pub use init_mask::{InitChunk, InitChunkIter};
 
+/// Functionality required for the bytes of an `Allocation`.
+pub trait AllocBytes:
+    Clone
+    + fmt::Debug
+    + Eq
+    + PartialEq
+    + Hash
+    + Deref<Target = [u8]>
+    + DerefMut<Target = [u8]>
+{
+    /// Adjust the bytes to the specified alignment -- by default, this is a no-op.
+    fn adjust_to_align(self, _align: Align) -> Self;
+
+    /// Create an `AllocBytes` from a slice of `u8`.
+    fn from_bytes<'a>(slice: impl Into<Cow<'a, [u8]>>, _align: Align) -> Self;
+
+    /// Create a zeroed `AllocBytes` of the specified size and alignment;
+    /// call the callback error handler if there is an error in allocating the memory.
+    fn zeroed<'tcx, F: Fn() -> InterpError<'tcx>>(
+        size: Size,
+        _align: Align,
+        handle_alloc_fail: F,
+    ) -> Result<Self, InterpError<'tcx>>;
+}
+
+// Default `bytes` for `Allocation` is a `Box<[u8]>`.
+impl AllocBytes for Box<[u8]> {
+    fn adjust_to_align(self, _align: Align) -> Self {
+        self
+    }
+
+    fn from_bytes<'a>(slice: impl Into<Cow<'a, [u8]>>, _align: Align) -> Self {
+        Box::<[u8]>::from(slice.into())
+    }
+
+    fn zeroed<'tcx, F: Fn() -> InterpError<'tcx>>(
+        size: Size,
+        _align: Align,
+        handle_alloc_fail: F,
+    ) -> Result<Self, InterpError<'tcx>> {
+        let bytes = Box::<[u8]>::try_new_zeroed_slice(size.bytes_usize())
+            .map_err(|_| handle_alloc_fail())?;
+        // SAFETY: the box was zero-allocated, which is a valid initial value for Box<[u8]>
+        let bytes = unsafe { bytes.assume_init() };
+        Ok(bytes)
+    }
```

Comment:
I think at this point we could just make this function return `Option` and leave the `None`/error handling to the caller entirely instead of passing in an error handler.

Opinions on that? Would it be useful for other implementors of `AllocBytes` to be able to have access to `handle_alloc_fail`?

Reply:
I think making it return an `Option` would be reasonable. I don't have any history with the interpreter code, but in terms of having access to handle_alloc_fail - I don't imagine so? It doesn't get passed anything, and it's an `Fn`, so using it to track state isn't really intended.

Reply:
Ok, let's switch to `Option`, or even better `Result<Self, ()>`, so that the callers can just `map_err`

Reply:
`Option` would let you do `ok_or` and not require taking a `()` param - do you have a strong preference for either? I personally lean towards `Option` for this case, but I can use `Result` if you prefer.

Reply:
No strong preference, let's go with `Option`